src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -614,20 +757,151 @@ func (s *fileSystemSource) readNextEntryAndIndex(
 	return err
 }
 
+func (s *fileSystemSource) incrementalBootstrapIndexSegment(
+	ns namespace.Metadata,
+	requestedRanges result.ShardTimeRanges,
+	runResult *runResult,
+) error {
+	// If we're performing an index run in incremental mode
+	// determine if we covered a full block exactly (which should
+	// occur since we always group readers by block size)
+	min, max := requestedRanges.MinMax()
+	blockSize := ns.Options().IndexOptions().BlockSize()
+	blockStart := min.Truncate(blockSize)
+
+	shards := make(map[uint32]struct{})
+	expectedRanges := make(result.ShardTimeRanges, len(requestedRanges))
+	for shard := range requestedRanges {
+		shards[shard] = struct{}{}
+		expectedRanges[shard] = xtime.Ranges{}.AddRange(xtime.Range{
+			Start: blockStart,
+			End:   blockStart.Add(blockSize),
+		})
+	}
+
+	indexResults := runResult.index.IndexResults()
+	indexBlock, ok := indexResults[xtime.ToUnixNano(blockStart)]
+	if !ok || !indexBlockHasSingleMutableSegment(indexBlock) {
+		return fmt.Errorf("missing single mutable segment")
+	}
+
+	mutableSegment := indexBlock.Segments()[0].(segment.MutableSegment)
+	fulfilled := indexBlock.Fulfilled()
+
+	// Check that completely fulfilled all shards for the block
+	// and we didn't bootstrap any more/less
+	requireFulfilled := expectedRanges.Copy()
+	requireFulfilled.Subtract(fulfilled)
+	exactStartEnd := min.Equal(blockStart) && max.Equal(blockStart.Add(blockSize))
+	if !exactStartEnd || !requireFulfilled.IsEmpty() {
+		return fmt.Errorf(
+			"incremental fs index bootstrap invalid ranges to persist: "+
+				"expected=%v, actual=%v, fulfilled=%v",
+			expectedRanges.String(),
+			requestedRanges.String(),
+			fulfilled.String())
+	}
+
+	// NB(r): Need to get an exclusive lock to actually write the segment out
```

Comment:
hm does this need to be done in `storage/flush.go` too? 

and do we pass the same persist manager instance to both the bootstrappers and the storage code?